### PR TITLE
Add support for long sleeps in JS scheduler

### LINKED ISSUE #2388: JS `TimeoutOverflowWarning`
Reported by @tpolecat when trying to set a timeout to `Long.MaxValue.ns`

```
(node:7170) TimeoutOverflowWarning: 9223372036854 does not fit into a 32-bit signed integer.
Timeout duration was set to 1.
```

Apparently the maximum timeout on JS is 2<sup>31</sup>-1 ms. See https://stackoverflow.com/questions/12633405/what-is-the-maximum-delay-for-setinterval

One workaround from here: https://stackoverflow.com/a/62002839/1642693

Essentially, to schedule a task that schedules another task and so on, such that the delays add up to the desired duration.

At the very least, we should round down to the maximum possible value instead of to 1 :)

### PR DESCRIPTION
**Pull request by armanbilge**
Fixes #2388. Also removes a duplicate (?) implementation of the default scheduler.

### CODE REVIEW

File: core/js/src/main/scala/cats/effect/unsafe/SchedulerCompanionPlatform.scala [hunk 1]
```diff
@@ -16,17 +16,24 @@
...
 
 private[unsafe] abstract class SchedulerCompanionPlatform { this: Scheduler.type =>
   def createDefaultScheduler(): (Scheduler, () => Unit) =
     (
       new Scheduler {
-        def sleep(delay: FiniteDuration, task: Runnable): Runnable = {
-          val handle = timers.setTimeout(delay)(task.run())
-          () => timers.clearTimeout(handle)
-        }
+        private[this] val maxTimeout = Int.MaxValue.millis
```
Comment (vasilmkd) [inline core/js/src/main/scala/cats/effect/unsafe/SchedulerCompanionPlatform.scala]: Should we move this on the companion object itself? Above in the class body.

File: core/js/src/main/scala/cats/effect/unsafe/SchedulerCompanionPlatform.scala [hunk 2]
```diff
@@ -16,17 +16,24 @@
...
+
+        def sleep(delay: FiniteDuration, task: Runnable): Runnable =
+          if (delay <= maxTimeout) {
+            val handle = timers.setTimeout(delay)(task.run())
+            () => timers.clearTimeout(handle)
+          } else {
+            var cancel: Runnable = () => ()
+            cancel = sleep(maxTimeout, () => cancel = sleep(delay - maxTimeout, task))
```
Comment (vasilmkd) [inline core/js/src/main/scala/cats/effect/unsafe/SchedulerCompanionPlatform.scala:35]: Sorry, can you explain exactly what this line does?

Comment (armanbilge) [reply core/js/src/main/scala/cats/effect/unsafe/SchedulerCompanionPlatform.scala:35]: We create a mutable reference to store the cancellation `Runnable` (initialized to a dummy). Then we schedule a task to occur after the max natively supported timeout and store its cancellation `Runnable`. The scheduled task itself schedules the target task with the remaining sleep time, and updates the reference with the new cancellation `Runnable`.

If that makes sense 😅

Comment (vasilmkd) [reply core/js/src/main/scala/cats/effect/unsafe/SchedulerCompanionPlatform.scala:35]: That's just awesome. 😆

**Review by vasilmkd — APPROVED** [review]

File: core/js/src/main/scala/cats/effect/unsafe/SchedulerCompanionPlatform.scala (hunk 2 above)
Comment (djspiewak) [reply core/js/src/main/scala/cats/effect/unsafe/SchedulerCompanionPlatform.scala:35]: …holy shit

**Review by djspiewak — APPROVED** [review]

**Pull request merged**